### PR TITLE
use path.sep to maintain cross-platform

### PR DESCRIPTION
**Pull request by kellyselden**
Fixes the remaining windows issues. The basic idea is that we accept any path format (/ or \\), but output paths in your OS's format, just like node's native path.join, path.normalize, etc.

Hopefully this can unblock the ember-cli 1.13.9 release. This PR may generate some discussion, as I think @stefanpenner may disagree with me. In my opinion, this is ready to merge.

cc @raytiley @felixrieseberg curious to hear other windows opinions if I'm barking up the wrong tree or not :stuck_out_tongue:

### CODE REVIEW

**Comment by stefanpenner** [comment]
I think I would prefer the internals normalize to posix, and consistently produce posix paths.

This matches how tools like miniglob work, and minimize interpath.sep mixing.

**Comment by joliss** [comment]
> I think I would prefer the internals normalize input to posix, and consistently produce posix paths.

I'd prefer this as well - I think we use forward slashes on Windows in most places throughout the Broccoli ecosystem. This is quite useful sometimes, and I'm trying to remember where exactly... Tests are one example. I think there are more examples, but can't come up with any right now.

@kellyselden, do you want to elaborate on the issue that motivated this PR?

**Comment by stefanpenner** [comment]
@joliss there is a legit issue, but it can be solved with consistent posix output.

**Comment by felixrieseberg** [comment]
Agree with @stefanpenner here. Having one consistent layer internally (that maps to the right path when talking to Windows) would probably make it less likely for someone to mess up in the future.

**Comment by kellyselden** [comment]
> do you want to elaborate on the issue that motivated this PR?

The problem jumped out in a test from broccoli-funnel:
```js
var expected = [
  [ path.join('__input_path__', 'subdir1/subsubdir1/foo.png'),
    path.join('__output_path__', 'foo/subdir1/subsubdir1/foo.png'),
    'subdir1/subsubdir1/foo.png' ],
  [ path.join('__input_path__', 'subdir1/subsubdir2/some.js'),
    path.join('__output_path__', 'foo/subdir1/subsubdir2/some.js'),
    'subdir1/subsubdir2/some.js' ]
];
```
In order to make this pass on Windows, one would have to change to:
```js
var expected = [
  [ path.join('__input_path__', 'subdir1/subsubdir1/foo.png').replace(/\\/g, '/'),
    path.join('__output_path__', 'foo/subdir1/subsubdir1/foo.png').replace(/\\/g, '/'),
    'subdir1/subsubdir1/foo.png' ],
  [ path.join('__input_path__', 'subdir1/subsubdir2/some.js').replace(/\\/g, '/'),
    path.join('__output_path__', 'foo/subdir1/subsubdir2/some.js').replace(/\\/g, '/'),
    'subdir1/subsubdir2/some.js' ]
];
```
This isn't very clean, and prone to failure IMO. My solution would be friendly to `path.join`'s normalizing since you're always working with normalized output. But, you would still have to adjust if you hardcode a Unix path:
```js
var expected = [
  [ path.join('__input_path__', 'subdir1/subsubdir1/foo.png'),
    path.join('__output_path__', 'foo/subdir1/subsubdir1/foo.png'),
    path.normalize('subdir1/subsubdir1/foo.png') ],
  [ path.join('__input_path__', 'subdir1/subsubdir2/some.js'),
    path.join('__output_path__', 'foo/subdir1/subsubdir2/some.js'),
    path.normalize('subdir1/subsubdir2/some.js') ]
];
```

> Having one consistent layer internally (that maps to the right path when talking to Windows)

@felixrieseberg I agree consistency in the internals, but I'm referring to output being the format you expect in Windows. Do you still disagree with me?

**Comment by felixrieseberg** [comment]
I totally agree with your assessment (didn't even disagree with the original solution). I just *also* happen to agree with Stefan, because people tend to forget to use `path` in modules and such - if Stefan has plans to handle all paths in Ember Cli (and to normalize internally), we could ensure that all paths are normalized when handed over to third party systems. This would have to use the same code you're implementing here, btw.

Long story short: This PR makes a lot of sense. It'd *also* make sense for Ember Cli to take ownership of path matters, given that we have a bunch of devs that tend to forget why Node has a `path` module.

**Comment by joliss** [comment]
One argument in favor of using forward slashes everywhere is this: Build definitions (`Brocfile.js` files) need to work across platforms, so we'll certainly be using forward-slashes there. So the best we'd be able to do is "use forward slashes as input, and output backslashes". That doesn't seem much of an improvement.

As for the `path.join` issue: I mostly use `path1 + '/' + path2` instead. This works across platforms. The one major difference to be aware of is that `path.join('', '') === '.'`, so you probably want to make sure that `path1` isn't the empty string.

The string concatenation may look weird at first, but I'll argue that it's actually better than `path.join`:

(1) It's much faster. `path.join` is surprisingly slow. I improved it some in https://github.com/nodejs/node-v0.x-archive/pull/6929, but it's still slow enough to matter in hot loops.

(2) It avoids the [`path.normalize`](https://nodejs.org/api/path.html#path_path_normalize_p) call implicitly done by `path.join`. To my mind, at best `path.normalize` function trades CPU cycles for a purely cosmetic change (that's mostly a no-op anyway). At worst, is actively harmful, because `foo/symlinkedDir/../bar` is *not* generally equivalent to `foo/bar`. So not normalizing is actually a feature, not a bug.

**Comment by kellyselden** [comment]
@joliss I agree with all you said. I think though people writing normal, unoptimized code using these libs are going to fall into the trap of `path` functions altering their paths in an inadvertent way. Evidence is in the tests from broccoli-funnel. Someone used a `path.join` and unknowingly introduced path inconsistencies. Training people to not use core node libs might be an interesting battle, but I could be wrong.

All that being said, I just wanted to make the issue visible, and If the consensus drives us to avoid normalizing, I can go down that path instead.

I do think it places the trickiness in the consumers code though, rather than putting the trickiness in our code.

**Comment by stefanpenner** [comment]
> I do think it places the trickiness in the consumers code though, rather than putting the trickiness in our code.

I'm fine with this trade-off

**Comment by joliss** [comment]
@kellyselden At least as test suites are concerned, I think the argument actually falls squarely in favor of using forward slashes everywhere. If we use backslashes, we'll still get bitten, e.g. by string literals (`expect(outputPath).to.equal('base_dir/out')`), but now we also have to resort to normalization code like [normalize-path-for-regex.js](https://github.com/joliss/node-walk-sync/pull/18/files#diff-48cac6e37ec36b883b64a3413596dd8cR4). I see path literals a lot in test code, so this makes test code more complex and harder to write.

**Comment by stefanpenner** [comment]
I believe https://github.com/joliss/node-walk-sync/pull/19 should do the trick

**Comment by joliss** [comment]
#19 looks like the best solution to me. @kellyselden, does #19 address your concerns adequately?

**Comment by kellyselden** [comment]
Thanks all for your input.
